### ISSUE TITLE
[Feature Request] OSA script (AppleScript & JavaScript) support

### ISSUE THREAD

**Issue by tzhao11** [lhc70000/iina#316]
Both Quicktime and VLC support automation via AppleScript and JavaScript.

**Comment by Henrietta1989** [comment]
+1 for AppleScript support. At least allow AppleScript to get current position of player. Returns "10" if video is playing or paused currently at 10 second.

**Comment by lawder** [comment]
+1 for AppleScript support.

**Comment by winray** [comment]
+1

**Comment by theLMGN** [comment]
@lawder @winray 
![image](https://user-images.githubusercontent.com/12257112/59933181-d4c4dd00-9440-11e9-9ef0-eddef2b20816.png)

Click this instead of creating useless +1 comments

**Comment by ideologysec** [comment]
Any updates on this? Would be great to replace VLC completely with IINA; as it is, I still keep VLC for the sole purpose of opening network streams via AppleScript from a browser.

**Comment by saagarjha** [comment]
No progress, but we do have browser extensions which might do what you want.

**Comment by theLMGN** [comment]
They can only play media. My usecase involves seeing what's playing

**Comment by saagarjha** [comment]
Right, doing any actual AppleScript-y things is going to require OSA support.

**Comment by theLMGN** [comment]
Also, @ideologysec, have you tried the following AppleScript?

`open location "iina://open?url=http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4"`

**Comment by Wevah** [comment]
I've done some preliminary work on this at (https://github.com/Wevah/iina/tree/applescript).

**Comment by lhc70000** [comment]
Hi @Wevah,

Thanks for your work! Would you mind opening a pull request? Although it's unfinished, I think we can see the diff more clearly and discuss any further problems there.

I'm working on the Javascript plugin system at https://github.com/iina/iina/tree/plugin-system. Perhaps some work in this branch can be reused.

**Comment by Wevah** [comment]
Sure, I can do that. I’m not completely sure about how I separated some things out, but I guess that’s what review is for.

**Comment by varenc** [comment]
@Wevah psyched to see your branch! I can't offer any serious help on the IINA implementation details, but there's anything simpler that'd still be useful for you let me know (like trying to bug test your new AppleScript interface?)

**Comment by Wevah** [comment]
Ah! I put this off because I got knee deep in rewriting a Punycode/IDNA thing to be more accurate/for NNW.

**Comment by shinepixel** [comment]
Any update regarding this topic? 
I wrote a script to seek backward/forward some seconds with media keys in my active music player (iTunes/Spotify/VLC). Unfortunately my favorite player is IINA and till now I couldn't come up with an idea how to control it.

**Comment by Wevah** [comment]
I'll see if I can get the branch up to date.

**Comment by Wevah** [comment]
Well, that was an easy rebase.

IIRC, a lot of the functionality is ready, but could use a review for style or other issues. I did my best to keep the type codes for commands the same as exist in Music.app or QuickTime Player (if they exist in those apps).

Stuff that's missing can always be added later.

**Comment by shinepixel** [comment]
So it is ready for testing? Where can I find it?

**Comment by Wevah** [comment]
> So it is ready for testing? Where can I find it?

There's public version built, but it's all up on the branch in the PR.

(Also feel free to assign this to me since I can't!)

**Comment by 45gfg9** [comment]
I know it's been over 3 years, where are we getting now? It's frustrating that such a good native app lacks AppleScript support.

**Comment by winray** [comment]
我已收到，我会尽快给您回复。

**Comment by ryanwhitney** [comment]
Just re-iterating that this would still be great to have! I have an iOS app that acts as a remote control for media applications via AppleScript and the lack of support makes using it with IINA less reliable than others.
